typings/styled-components.d.ts

```diff
@@ -35,13 +33,14 @@ type Attrs<P, A extends Partial<P>, T> = {
 export interface StyledComponentClass<P, T, O = P> extends ComponentClass<ThemedOuterStyledProps<O, T>> {
   extend: ThemedStyledFunction<P, T, O>;
 
-  withComponent<K extends keyof JSX.IntrinsicElements>(tag: K): StyledComponentClass<JSX.IntrinsicElements[K], T, O>;
-  withComponent(element: ComponentClass<P>): StyledComponentClass<P, T, O>;
+  withComponent<K extends keyof JSX.IntrinsicElements>(tag: K): StyledComponentClass<JSX.IntrinsicElements[K], T, JSX.IntrinsicElements[K]>;
```

Comment:
The reason to reset `outerProps` to `JSX.IntrinsicElements[K]` is because that `O` is inferred from `StyledComponentClass` instance here, 
which is not the case for `.withComponent()` usage for `JSX.IntrinsicElements[K]`, you can read more detailed explanation in #1315.
For the use case with pre-defined `attrs`, I do not know if `withComponent` will inherit them from base component instance, and I do not think `attrs` should be inherited in this case either since we try to reuse the style not pre-defined `attrs` ?

Reply:
But we could just set `StyledComponentClass<JSX.IntrinsicElements[K], T>` here to omit extra typings.